### ISSUE TITLE
Correct way to package this in distributions

### ISSUE THREAD

**Issue by faulesocke** [rust-lang/rustup#1559]
What is the correct way to package rustup for shipping with linux distributions? When we just install the compiled `rustup-init` as `/usr/bin/rustup` it does not seem to put anything into `~/.cargo/bin`. However, when we install it as `/usr/bin/rustup-init`, the users will just have to run `rustup-init` every time the package gets updated in order to get the correct version installed into their `~/.cargo/bin`. So what is the correct way to do things, i.e. how can I get rustup to install _symlinks_ into `~/.cargo/bin` correctly (hardlinks won't probably work since it's a common pattern that `/home` and `/usr` reside on different file systems).

**Comment by kinnison** [comment]
The "correct" way would be for your package to set up all the hardlinks for `rustup` in `/usr/bin`.

In addition you should ensure that you build with the `no-self-update` feature enabled so that `rustup` won't try and replace itself which would otherwise go pretty wrong.

However I'd note that `rustup` is not really meant to be packaged by distributions, at least not right now, so your experience won't be optimal.

**Comment by Cogitri** [comment]
Distros have to package rustup in case they support arches rustup doesn't (yet) support though. In Void Linux and Alpine Linux I've done it by patching rustup to make symlinks in `$HOME/.cargo/bin` instead of hardlinks to allow the package manager to update rustup which works for now.

**Comment by kinnison** [comment]
@Cogitri If there are platforms for which `rustc` and `cargo` are built and for which we don't provide a `rustup` then it's worth letting us know (in independent issues) so that we can look at providing such builds.

**Comment by Cogitri** [comment]
Yup, but e.g. when rust started supporting x86_64-unkown-linux-musl rustup didn't immediately catch up to that IIRC (not sure if it supports it now). Additionally some people prefer their package manager keeping track of as much as possible (without sacrificing the flexibility rustup has).

**Comment by kinnison** [comment]
Fair enough.  We do support musl now I think, but I appreciate there's a delay.

**Comment by faulesocke** [comment]
> However I'd note that `rustup` is not really meant to be packaged by distributions, at least not right now, so your experience won't be optimal.

The symlink-patch from @Cogitri works quite nice to give an optimal experience.
